core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -146,16 +157,15 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
       }
 
-      val leaderAndIsrResponse =
-        if (authorize(request.session, ClusterAction, Resource.ClusterResource)) {
-          val result = replicaManager.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest, onLeadershipChange)
-          new LeaderAndIsrResponse(result.error, result.responseMap.asJava)
-        } else {
-          val result = leaderAndIsrRequest.partitionStates.asScala.keys.map((_, Errors.CLUSTER_AUTHORIZATION_FAILED)).toMap
-          new LeaderAndIsrResponse(Errors.CLUSTER_AUTHORIZATION_FAILED, result.asJava)
-        }
-
-      requestChannel.sendResponse(new Response(request, leaderAndIsrResponse))
+      if (authorize(request.session, ClusterAction, Resource.ClusterResource)) {
+        val result = replicaManager.becomeLeaderOrFollower(correlationId, leaderAndIsrRequest, onLeadershipChange)
+        val leaderAndIsrResponse = new LeaderAndIsrResponse(result.error, result.responseMap.asJava)
+        sendResponseExemptThrottle(request, new Response(request, leaderAndIsrResponse))
+      } else {
+        val result = leaderAndIsrRequest.partitionStates.asScala.keys.map((_, Errors.CLUSTER_AUTHORIZATION_FAILED)).toMap
+        def createResponse(throttleTimeMs: Int): AbstractResponse = new LeaderAndIsrResponse(Errors.CLUSTER_AUTHORIZATION_FAILED, result.asJava)
+        sendResponseMaybeThrottle(request, createResponse)
```

Comment:
@junrao Thank you for the review. 

It will be nice to handle throttling in a single place. However, handling of all new requests need to be aware of throttling, so that they add the throttling time to the response. The bigger issue is the exclusions. We need to authorize in a central place for `ClusterAction`. And worse, we need to handle `Produce` differently since we don't want to throttle until after the request is processed and the memory can be released. We probably want to throttle later for `Fetch` as well since we are relying on the bandwidth throttle time to reduce the request throttle time. We will need to record two more timestamps to take into account time spent before throttle in some cases and after throttle in others. And as you mentioned, we still want the request to be processed after throttling on the request handler thread. Taking all that into account, I am not  sure it is worthwhile to restructure the code to centralize the throttling logic.

I have updated `RequestQuotaTest` to ensure that unauthorized requests of all types are throttled. And also to check that all requests not explicitly excluded are throttled and return throttle time in response. This should catch any missing throttling in new requests.

Let me know if this is sufficient or whether I should try out the centralized approach.